Functions/ImageProcessing/xASL_im_ASLSubtractionAveraging.m

```diff
@@ -0,0 +1,317 @@
+    % which is deleted right after we reload its NIfTI as image volumes.
+    
+    if numel(path_ASL4D)~=1 && ~isnumeric(path_ASL4D) && numel(path_ASL4D)<1000
+        % here we detect a path
+        [Fpath, Ffile] = xASL_fileparts(path_ASL4D);
+        sideCarMat = fullfile(Fpath, [Ffile '.mat']);
+    
+        if exist(sideCarMat, 'file')
+            % we check if the mat-sidecar exists, with the motion estimation parameters
+            % If this file exists, we assume that the original ASL4D input file has not been
+            % interpolated yet (and the motion estimation has not been applied yet)
+    
+            % Apple motion correction & resample
+            if nVolumes>1
+                for iS=1:nVolumes
+                    matlabbatch{1}.spm.spatial.realign.write.data{iS,1} = [path_ASL4D ',' num2str(iS)];
+                end
+            
+                matlabbatch{1}.spm.spatial.realign.write.roptions.which     = [2 0];
+                matlabbatch{1}.spm.spatial.realign.write.roptions.interp    = 4;
+                matlabbatch{1}.spm.spatial.realign.write.roptions.wrap      = [0 0 0];
+                matlabbatch{1}.spm.spatial.realign.write.roptions.mask      = 1;
+                matlabbatch{1}.spm.spatial.realign.write.roptions.prefix    = 'r';
+            
+                spm_jobman('run',matlabbatch); % this applies the motion correction in native space
+                
+                path_rASL4D = fullfile(Fpath, ['r' Ffile '.nii']);
+                
+                % RELOAD ASL4D.nii
+                % Note that we only reload ASL4D.nii here if we have motion corrected timeseries.
+                % If path_ASL4D was not a path, or if sideCarMat didn't exist (motion estimation was not performed)
+                % or if nVolumes==1 (we did not have multiple ASL volumes) there is no reason to resample and we keep using
+                % the already above loaded ASL4D.nii
+                ASL4D = xASL_io_Nifti2Im(path_rASL4D);
+                xASL_delete(path_rASL4D);
+            end
+        end
+    end
+
+
+    %% 2B. Apply time decoding if needed
+    if x.modules.asl.bTimeEncoded
+        %% #997 Check that the motion correction of Hadamard is correctly performed here,
+        % should we use path_rASL4D from here onwards, which defaults to path_ASL4D?
+
+        % Decoding of TimeEncoded data - it outputs the decoded image as well as the parameters
+        [PWI4D, Control4D, x.Q] = xASL_quant_HadamardDecoding(path_ASL4D, x.Q);
+	    nVolumes = size(PWI4D, 4);
+
+
+    %% 2C. Non-time encoded subtraction
+    else
+        if nVolumes>1 && ~x.modules.asl.bContainsDeltaM
+            % Paired subtraction
+            [Control4D, Label4D] = xASL_quant_GetControlLabelOrder(ASL4D);
+            PWI4D = Control4D - Label4D;
+            
+            % Skip every other value in the vectors as they were stored for both control and label images 
+            x.Q.EchoTime_PWI4D = x.Q.EchoTime(1:2:end);
```

Comment:
Good point, you mean that xASL_quant_GetControlLabelOrder is equivalent to xASL_quant_HadamardDecoding, in the sense that it outputs Control4D and deltaM/PWI4D, as well as the changed x.Q?

Do you want to do this, since you're also doing this for xASL_quant_HadamardDecoding, so it's easier for you to standardize this? 

We could even consider making this a single function xASL_im_ControlLabelDecoding that will do/manage this for all types of ASL sequences.

Reply:
Will do.

Reply:
Done in 4e989679
We can create a single function if you want. But we should only do that in #1543 to avoid further conflicts.